### PR TITLE
Building OSL on Windows OS

### PR DESCRIPTION
**Pull request by madoodia**
## Description

This PR will add a method of building OSL on Windows OS.

The whole process is explained in an `md` file in `doc/build_install/windows` directory

## Tests

I've tested the whole process on Windows 10 and OSL is installed successfully

## Checklist:

- [x] I have read the [contribution guidelines](../CONTRIBUTING.md).
- [x] I have previously submitted a [Contributor License Agreement](http://opensource.imageworks.com/cla/).
- [x] I have updated the documentation, if applicable.
- [ ] I have ensured that the change is tested somewhere in the testsuite (adding new test cases if necessary).
- [x] My code follows the prevailing code style of this project.

### CODE REVIEW

**Comment by mprater** [comment]
Thanks Reza. I wonder if you could convert the .odt file to .md or .rst? We'll want documentation to use either of those formats in the future.

I used .odt as a quicky container for content I grabbed from elsewhere.

**Comment by madoodia** [comment]
> Thanks Reza. I wonder if you could convert the .odt file to .md or .rst? We'll want documentation to use either of those formats in the future.
> 
> I used .odt as a quicky container for content I grabbed from elsewhere.

Sure, I'll do that
Sorry

**Comment by mprater** [comment]
It appears that many of these files contain copyright notices in their comments. Those ought to be removed; or replaced with:

Copyright Contributors to the Open Shading Language project.

**Comment by madoodia** [comment]
> Copyright Contributors to the Open Shading Language project.

I see
so errors are because of them
I'll remove them
thanks for letting me know

**Review by mprater — APPROVED** [review]
I have no way to test any of this, but what's here looks okay to me. Thanks Reza!

File: doc/build_install/windows/Readme.md [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+4. Running
+5. Development in Windows
+
+---
+
+### **1. Introduction**
+
+This document is prepared for those who have problem in building OSL in Windows OS, this is completely custom setup. And it is developing to make it compatible with all situations.
+
+(`Disclaimer: This Process is completely Experimental and it worked on my system, maybe you get error with changing some configs, during this process). At a first glance it seems to be complicated, but it is very easy to do, bare with me :)
```
Comment (lgritz) [inline doc/build_install/windows/Readme.md:30]: I think you have an opening single quote here, that does not have a corresponding closing quote.

File: doc/build_install/windows/build_osl.py [hunk 2]
```diff
@@ -0,0 +1,2355 @@
+#
+# Copyright Contributors to the Open Shading Language project.
+# 
+# Based on USD build_scripts by Pixar Animation Studio
```
Comment (lgritz) [inline doc/build_install/windows/build_osl.py:4]: If this contains any code copied from USD, we need to reproduce their copyright and license information here to be in compliance with that project's Apache 2.0 license.

**Comment by lgritz** [comment]
Every file (if at all possible) in the project needs both of these lines somewhere (usually as comments at the very top of the file):

    Copyright Contributors to the Open Shading Language project.
    SPDX-License-Identifier: BSD-3-Clause

We use BSD-3-Clause for code files, but if the file is purely documentation, we instead use a Creative Commons license, so that would instead read:

    SPDX-License-Identifier: CC-BY-4.0

Any material "imported" from any other project -- that is, not written from scratch for the OSL project -- should have a comment that says (a) The copyright notice preferred by the project it came from, (b) the appropriate [SPDX identifier](https://spdx.org/licenses/) for the license used by that project, and (c) exactly which code that applies to (this whole file? a particular function or set of functions? just these three lines?).

Also the top-level file THIRD-PARTY.md details reproduces the licenses used by any third-party bits in our code base, and says which parts/projects are associated with which license.

**Comment by madoodia** [comment]
Thanks @lgritz 
I will edit files

File: doc/build_install/windows/Readme.md (hunk 1 above)
Comment (madoodia) [reply doc/build_install/windows/Readme.md]: I fixed this one

File: doc/build_install/windows/build_osl.py (hunk 2 above)
Comment (madoodia) [reply doc/build_install/windows/build_osl.py:7]: I've added copyright notice to all files
just I'm not sure about the Pixar USD reproduced copyright
please check the file

File: THIRD-PARTY.md [hunk 3]
```diff
@@ -7,46 +7,45 @@ into parts of OSL.
...
-* OpenEXR, Copyright (c) Contributors to the OpenEXR Project.
+- OpenEXR, Copyright (c) Contributors to the OpenEXR Project.
   https://github.com/AcademySoftwareFoundation/OpenEXR
 
   OSL replicates and modifies several files from the "Imath" portion of
   OpenEXR. These are all found in src/include/OSL/Imathx.
 
-* Qt, Copyright (c) The Qt Company Ltd.
+- Qt, Copyright (c) The Qt Company Ltd.
   http://doc.qt.io/
```
Comment (lgritz) [inline THIRD-PARTY.md:29]: Please revert all *changes* to THIRD-PARTY.md.

Don't change anything that's already there. Only add what you need for your particular change. If you think something else in this file is incorrect, let's handle that in a separate PR.

File: doc/build_install/windows/build_osl.py (hunk 2 above)
Comment (lgritz) [reply doc/build_install/windows/build_osl.py:7]: Here's how I think the header of this file should read:

```
# Copyright Contributors to the Open Shading Language project.
# SPDX-License-Identifier: BSD-3-Clause
#
# Portions of this file were derived from USD, which has the following notice:
# Copyright Pixar
# Apache 2.0 license with modifications: https://github.com/PixarAnimationStudios/USD/blob/release/LICENSE.txt
```

File: THIRD-PARTY.md [hunk 4]
```diff
@@ -93,21 +90,193 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
...
   you like with it. You may even remove all attributions,
   but of course I'd appreciate it if you kept my name somewhere.
 
-
--------------------------------------------------------------------------
+---
+
+                               Modified Apache 2.0 License
+
+TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION
```
Comment (lgritz) [inline THIRD-PARTY.md:109]: We need to say here WHICH parts of our code (and which codebases we derived from) use this reproduced license, just like in all the sections above. So something like:

Modified version of Apache-2.0

* Universal Scene Description
  Copyright Pixar.

  Portions of the build_osl.py were derived from scripts distributed with USD.

... then you reproduce the license ...

File: THIRD-PARTY.md (hunk 3 above)
Comment (madoodia) [reply THIRD-PARTY.md:29]: Sorry about that @lgritz 
these lines are automatically corrected by VSCode autoformat of `md` files
was not intentionally

File: THIRD-PARTY.md (hunk 4 above)
Comment (madoodia) [reply THIRD-PARTY.md:109]: @lgritz I copied this part from the USD file, lots of lines of build_usd.py is edited. don't know how to mention all the edited lines tbh.

Comment (lgritz) [reply THIRD-PARTY.md]: Do just like I said above -- insert those few lines right above the license which you reproduced.

File: THIRD-PARTY.md (hunk 3 above)
Comment (lgritz) [reply THIRD-PARTY.md]: thanks

**Comment by mprater** [comment]
Merging now. Will make English language edits at a later time.

**Comment by madoodia** [comment]
> Merging now. Will make English language edits at a later time.

Yes, it needs lots of edits.
thanks though @mprater

**Pull request merged**